File: python/scrabble-score/scrabble_score.py
scores = {}
letter_scores = [
    (['a', 'e', 'i', 'o', 'u', 'l', 'n', 'r', 's', 't'], 1),
    (['d', 'g'], 2),
    (['b', 'c', 'm', 'p'], 3),
    (['f', 'h', 'v', 'w', 'y'], 4),
    (['k'], 5),
    (['j', 'x'], 8),
    (['q', 'z'], 10)
]

for ls in letter_scores:
    for l in ls[0]:
        scores[l] = ls[1]

def score(word):
    sum = 0
    for c in word.lower():
        sum += scores[c]
    return sum

File: ruby/scrabble-score/scrabble_score.rb
class Scrabble
  def self.score(word)
    return 0 if word == nil
    word.chars.reduce(0) { |sum, char| sum + @@score_by_letter.fetch(char.upcase, 0) }
  end

  def self.populate_scores()
    compressed_scores = {
      1 => "AEIOULNRST",
      2 => "DG",
      3 => "BCMP",
      4 => "FHVWY",
      5 => "K",
      8 => "JX",
      10 => "QZ",
    }

    @@score_by_letter = {}
    compressed_scores.each { |score, letters| letters.chars.each { |letter| @@score_by_letter[letter] = score } }
  end

  attr_reader :score

  def initialize(word)
    Scrabble.populate_scores if defined?(@@score_by_letter) == nil
    @score = Scrabble.score(word)
  end
end
